waveorder/filter.py

```diff
@@ -0,0 +1,196 @@
+import itertools
+import numpy as np
+import torch
+
+
+def apply_filter_bank(
+    io_filter_bank: torch.Tensor,
+    i_input_array: torch.Tensor,
+) -> torch.Tensor:
+    """
+    Applies a filter bank to an input array.
+
+    io_filter_bank.shape must be smaller or equal to i_input_array.shape in all
+    dimensions. When io_filter_bank is smaller, it is effectively "stretched"
+    to apply the filter.
+
+    io_filter_bank is in "wrapped" format, i.e., the zero frequency is the
+    zeroth element.
+
+    i_input_array and io_filter_bank must have inverse sample spacing, i.e.,
+    is input_array contains samples spaced by dx, then io_filter_bank must
+    have extent 1/dx. Note that there is no need for io_filter_bank to have
+    sample spacing 1/(n*dx) because io_filter_bank will be stretched.
+
+    Parameters
+    ----------
+    io_filter_bank : torch.Tensor
+        The filter bank to be applied in the frequency domain.
+        The spatial extent of io_filter_bank must be 1/dx, where dx is the
+        sample spacing of i_input_array.
+
+        Leading dimensions are the input and output dimensions.
+        io_filter_bank.shape[:2] == (i, o)
+
+        Trailing dimensions are spatial frequency dimensions.
+        io_filter_bank.shape[2:] == (Z', Y', X') or (Y', X')
+
+    i_input_array : torch.Tensor
+        The input array with sample spacing dx to be filtered.
+
+        Leading dimension is the input dimension, matching the filter bank.
+        i_input_array.shape[0] == i
+
+        Trailing dimensions are spatial dimensions.
+        i_input_array.shape[1:] == (Z, Y, X) or (Y, X)
+
+    Returns
+    -------
+    torch.Tensor
+        The filtered output array with the same shape and dtype as input_array.
+    """
+    
+    # Ensure all dimensions of transfer_function are smaller than or equal to input_array
+    if any(
+        t > i
+        for t, i in zip(io_filter_bank.shape[2:], i_input_array.shape[1:])
+    ):
+        raise ValueError(
+            "All spatial dimensions of io_filter_bank must be <= i_input_array."
+        )
+
+    # Ensure the number of spatial dimensions match
+    if io_filter_bank.ndim - i_input_array.ndim != 1:
+        raise ValueError(
+            "io_filter_bank and i_input_array must have the same number of spatial dimensions."
+        )
+
+    # Ensure the input dimensions match
+    if io_filter_bank.shape[0] != i_input_array.shape[0]:
+        raise ValueError(
+            "io_filter_bank.shape[0] and i_input_array.shape[0] must be the same."
+        )
+
+    I, O = io_filter_bank.shape[:2]
```

Comment:
What is `I` and `O`? Is it possible to use lower case readable words for these variables?

Reply:
changed `I` -> `num_input_channels` and `O` -> `num_output_channels`. Thanks!